vm/systemSmartContracts/auction.go

```diff
@@ -0,0 +1,694 @@
+	blsKeys := registrationData.BlsPubKeys
+	if len(args.Arguments) > 0 {
+		for _, argKey := range args.Arguments {
+			found := false
+			for _, blsKey := range blsKeys {
+				if bytes.Equal(argKey, blsKey) {
+					found = true
+					break
+				}
+			}
+
+			if !found {
+				log.Debug("not allowed to unbound for bls key which is not under validator")
+				return nil, vm.ErrBLSPublicKeyMissmatch
+			}
+		}
+
+		blsKeys = args.Arguments
+	}
+
+	return blsKeys, nil
+}
+
+func (s *stakingAuctionSC) unBound(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	registrationData, err := s.getRegistrationData(args.CallerAddr)
+	if err != nil {
+		return vmcommon.UserError
+	}
+
+	blsKeys, err := getBLSPublicKeys(registrationData, args)
+	if err != nil {
+		return vmcommon.UserError
+	}
+
+	totalUnBound := big.NewInt(0)
+	for _, blsKey := range blsKeys {
+		stakedData, err := s.getStakedData(blsKey)
+		if err != nil || len(stakedData.RewardAddress) == 0 {
+			log.Debug("bls key was not staked")
+			return vmcommon.UserError
+		}
+
+		if stakedData.Staked || stakedData.UnStakedNonce <= stakedData.StartNonce {
+			log.Debug("unBound is not possible for address which is staked or is not in unbound period")
+			return vmcommon.UserError
+		}
+
+		currentNonce := s.eei.BlockChainHook().CurrentNonce()
+		if currentNonce-stakedData.UnStakedNonce < s.unBoundPeriod {
+			log.Debug("unBound is not possible for address because unbound period did not pass")
+			return vmcommon.UserError
+		}
+
+		s.eei.SetStorage(blsKey, nil)
+		config := s.getConfig(stakedData.UnStakedEpoch)
+
+		ownerAddress := s.eei.GetStorage([]byte(ownerKey))
+		err = s.eei.Transfer(args.CallerAddr, ownerAddress, config.NodePrice, nil)
+		if err != nil {
+			log.Debug("transfer error on finalizeUnStake function",
+				"error", err.Error(),
+			)
+			return vmcommon.UserError
+		}
+
+		_ = totalUnBound.Add(totalUnBound, config.NodePrice)
+	}
+
+	if registrationData.BlockedStake.Cmp(totalUnBound) < 0 {
+		log.Debug("too much to unbound, not enough total stake")
+		return vmcommon.UserError
+	}
+
+	_ = registrationData.BlockedStake.Sub(registrationData.BlockedStake, totalUnBound)
+	_ = registrationData.TotalStakeValue.Sub(registrationData.TotalStakeValue, totalUnBound)
+
+	zero := big.NewInt(0)
+	if registrationData.BlockedStake.Cmp(zero) == 0 && registrationData.TotalStakeValue.Cmp(zero) == 0 {
+		s.eei.SetStorage(args.CallerAddr, nil)
+	} else {
+		err := s.saveRegistrationData(args.CallerAddr, registrationData)
+		if err != nil {
+			log.Debug("cannot save registration data change")
+			return vmcommon.UserError
+		}
+	}
+
+	return vmcommon.Ok
+}
+
+func (s *stakingAuctionSC) claim(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	registrationData, err := s.getRegistrationData(args.CallerAddr)
+	if err != nil {
+		return vmcommon.UserError
+	}
+
+	if len(registrationData.RewardAddress) == 0 {
```

Comment:
Same checks all over the place. Might move them inside getRegistrationData

Reply:
done.